textcursor.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2015 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gxui
+
+import "github.com/google/gxui/interval"
+
+// TextCursor represents a text position and selection. A selection is defined as
+// the pair of cursor indices sorted least to greatest. When transforming a
+// cursor with no selection, mark should equal index.
+type TextCursor struct {
```

Comment:
If I understand things correctly, `index` and `mark` are equivalent to the old `start` and `end` fields, except the carat is always displayed at `mark`, and `mark` can be before `index`. If this is true, `index` and `mark` are still essentially a half-closed interval. What's not clear is which endpoint is open and which is closed (I guess the smaller of the two is always the closed endpoint). It would be good to have this documented.

I'm not sold on the field names `index` and `mark`. Given that they're interval endpoints, how about `From` and `To`? I'm not completely opposed to `mark`, but `index` just doesn't seem to work nicely as a counterpart for me. I'm open to other suggestions here. Maybe `Carat` and `Base`? I'll stick with `From` and `To` for the rest of my comments.

I'm not sure it's worth bringing in the concept of selection into the `TextCursor` documentation - it's rather confusing as it guarantees order (unlike `TextCursor`), and is already well explained below.

My stab at documentation (based on my understanding of the code) is as follows (I'm a visual guy):
```go
// TextCursor represents an interval of runes, where From and To can be 
// considered as the indices to the gaps between runes. 
// To can be less than, equal or greater to From.
// When the TextCursor is used as a TextBox selection, the carat should always 
// be displayed at To.
// The TextCursor represents a single point in a string When the To and From fields are equal.
//
// For example, given the string "Hello world":
//
//      ┌   ┬   ┬   ┬   ┬   ┬   ┬   ┬   ┬   ┬   ┬   ┐
//        H   e   l   l   o       w   o   r   l   d  
//      └   ┴   ┴   ┴   ┴   ┴   ┴   ┴   ┴   ┴   ┴   ┘
//      ₀   ₁   ₂   ₃   ₄   ₅   ₆   ₇   ₈   ₉   ₁₀  ₁₁
//
//
// TextCursor{ From: 0, To: 5 } represents:
//
//      ┌   ┬   ┬   ┬   ┬   ╥   ┬   ┬   ┬   ┬   ┬   ┐
//      │ H   e   l   l   o ║     w   o   r   l   d  
//      └   ┴   ┴   ┴   ┴   ╨   ┴   ┴   ┴   ┴   ┴   ┘
//      ├───────────────────╢ 
//      ₀                   ₅
//
//
// TextCursor{ From: 11, To: 6 } represents:
//
//      ┌   ┬   ┬   ┬   ┬   ┬   ╥   ┬   ┬   ┬   ┬   ┐
//        H   e   l   l   o     ║ w   o   r   l   d  
//      └   ┴   ┴   ┴   ┴   ┴   ╨   ┴   ┴   ┴   ┴   ┘
//                              ╟───────────────────┤
//                              ₆                   ₁₁
//
//
// TextCursor{ From: 5, To: 5 } represents:
//
//      ┌   ┬   ┬   ┬   ┬   ╥   ┬   ┬   ┬   ┬   ┬   ┐
//        H   e   l   l   o ║     w   o   r   l   d  
//      └   ┴   ┴   ┴   ┴   ╨   ┴   ┴   ┴   ┴   ┴   ┘
//                          ║
//                          ₅
```